opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/TracingHandlerInterceptor.java

```diff
@@ -0,0 +1,112 @@
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.servlet.ServletRequest;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.springframework.web.servlet.handler.HandlerInterceptorAdapter;
+
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.contrib.web.servlet.filter.HttpServletRequestExtractAdapter;
+import io.opentracing.propagation.Format;
+import io.opentracing.tag.Tags;
+
+/**
+ * Tracing handler interceptor for spring web. It creates a new span for an incoming request.
+ * Span context of created span can be accessed via {@link #serverSpanContext(ServletRequest)}.
+ *
+ * @author Pavol Loffay
+ */
+public class TracingHandlerInterceptor extends HandlerInterceptorAdapter {
+
+    /**
+     * Used as a name of {@link HttpServletRequest#setAttribute(String, Object)} to inject server span context
+     */
+    public static final String SERVER_SPAN_CONTEXT = TracingHandlerInterceptor.class.getName() + ".activeSpanContext";
+    private static final String ACTIVE_SPAN_WRAPPER = TracingHandlerInterceptor.class.getName() + ".activeSpan";
+
+    private Tracer tracer;
+    private List<SpanDecorator> decorators;
+
+    /**
+     * @param tracer tracer
+     * @param decorators span decorators
+     */
+    public TracingHandlerInterceptor(Tracer tracer, List<SpanDecorator> decorators) {
+        this.tracer = tracer;
+        this.decorators = new ArrayList<>(decorators);
+    }
+
+    @Override
+    public boolean preHandle(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse, Object handler)
+            throws Exception {
+
+        if (httpServletRequest.getAttribute(ACTIVE_SPAN_WRAPPER) == null &&
+                    isTraced(httpServletRequest, httpServletResponse, handler)) {
+
+            SpanContext parentContext = tracer.extract(Format.Builtin.HTTP_HEADERS,
+                    new HttpServletRequestExtractAdapter(httpServletRequest));
+
+            Span serverSpan = tracer.buildSpan(httpServletRequest.getMethod())
+                    .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_SERVER)
+                    .asChildOf(parentContext)
+                    .start();
+
+            for (SpanDecorator decorator: decorators) {
+                decorator.onPreHandle(httpServletRequest, handler, serverSpan);
+            }
+
+            // when an exception is thrown in controller `afterCompletion` is called twice, therefore using
+            // spanWrapper (span#finish()) is not idempotent by specification.
+            httpServletRequest.setAttribute(ACTIVE_SPAN_WRAPPER, new SpanWrapper(serverSpan, false));
+            httpServletRequest.setAttribute(SERVER_SPAN_CONTEXT, serverSpan.context());
+        }
+
+        return true;
+    }
+
+    @Override
+    public void afterCompletion(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse,
```

Comment:
Spring boot has a generic error handling and after "normal" processing it invokes `preHandler` and `afterCompletion` for error handling/view `/error`. I thought that it invokes only `afterCompletion`...

I see two possible solutions:
1. use span wrapper with idempotent finish
2. remove span from request attributes in `afterCompletion`. Then it would start new span for `/error`.
 
@bensigelman what do you think?